Job board

sophm256 opened the pull request:
### Creating a job post: These two pics are top and bottom of the same page
![job-post-form-1](https://user-images.githubusercontent.com/53249319/79089825-2b234500-7d0d-11ea-8557-a832f2ac058f.JPG)

![job-post-form-2](https://user-images.githubusercontent.com/53249319/79089831-2fe7f900-7d0d-11ea-8754-8790f19b9f3f.JPG)

### Page after you submit the job posting
![after-submit-job-post](https://user-images.githubusercontent.com/53249319/79089549-24480280-7d0c-11ea-947d-0ea81842b5a3.JPG)

### Django Admin Dashboard  showing a list of all job posting. Note that they have different statuses. 'status_change_date' is a read only field that shows up here (in the list). That field doesn't show up in the individual job post admin page because it causes an error so I took it out from there.
Every time an admin changes the 'status' field for a job post, this 'status_change_date' gets reset to the datetime when that change occurs. This 'status_change_date' field allows one to keep track of how much time has passed from the 'approval date' or if a post is granted an extension.

So far the 'status' can only be set manually by the admin. But in the future, it would be good to have the status change programmatically. For example, the status 'approved' would change to 'archived' after X number of days. 

![django-admin-list-of-job-posts](https://user-images.githubusercontent.com/53249319/79089585-46da1b80-7d0c-11ea-98db-dba9aa2b15fe.JPG)

### This is an example of what one job post looks like in the Django admin. Note that the last field is a boolean field that is the 'agree_to_terms' field. The label is invisible because I set 'verbose_name' to be empty string. I did this so that could have my own custom text with hyperlink in the job_post_list.hml' template, where I said "I have read and agree to the referral terms....etc".
![django-admin-one-job-post](https://user-images.githubusercontent.com/53249319/79089591-49d50c00-7d0c-11ea-8e9a-c8cc3a6d9945.JPG)

### The Job Post List page
![job-post-list](https://user-images.githubusercontent.com/53249319/79089677-96204c00-7d0c-11ea-8aae-7424d3470d43.JPG)

### Job Post Detail Page
![job-post-detail](https://user-images.githubusercontent.com/53249319/79089679-9882a600-7d0c-11ea-97d6-5b85e568d6e5.JPG)

sophm256 commented:
This is 2 pics are top and bottom of the same page. This is from the Django Admin. They are what a 'User' object and 'UserProfile' object look like. I have stored extra info about the 'contact' person (i.e. the user who is trying to post the job) in the 'UserProfile'. I added the field asking if they were an external recruiter. The other two fields were there previously.

![django-admin-what-user-object-and-profile-object-looks-like1](https://user-images.githubusercontent.com/53249319/79092108-688bd080-7d15-11ea-89f6-2915b1ad22cf.JPG)
![django-admin-what-user-object-and-profile-object-looks-like2](https://user-images.githubusercontent.com/53249319/79092109-6b86c100-7d15-11ea-9c00-decf47193898.JPG)

elmq0022 commented:
This looks really good from the pictures.  Will try to review soon.  Looks like you did a ton of work here.

sophm256 commented:
Thanks @elmq0022 :smile: . Thanks also for adding my reviewers.

sophm256 commented:
Hey @elmq0022 and @alysivji  , I made another commit that fixes some issues. For example, before the 'agree_to_terms' checkbox was not raising a validation error if the user didn't check it. 

Here are some pics that show the new changes. Changes are in orange. 
### Top Half of Create a Job Post Form
![after-change-create-job-post1-starred](https://user-images.githubusercontent.com/53249319/79173969-d5a77080-7dbe-11ea-8512-e7fa408f3355.JPG)
### Lower Half of Create a Job Post Form
![after-change-create-job-post2-starred](https://user-images.githubusercontent.com/53249319/79174134-61210180-7dbf-11ea-90b8-2c741c097d41.JPG)

### One Job Post Object in Django Admin
![after-change-one-job-post-starred](https://user-images.githubusercontent.com/53249319/79173995-f53e9900-7dbe-11ea-99d7-04a25a1a50f1.JPG)

elmq0022 commented on chipy_org/apps/job_board/models.py:
```diff
@@ -1,3 +1,55 @@
+        default=False, verbose_name="Is the company a sponsor of ChiPy?")
+
+    company_sponsor = models.ForeignKey(Sponsor, blank=True, null=True)
+
```
Can this just be called sponsor?   Doing so would make our templates more consistent and intuitive.

sophm256 replied:
Hey @elmq0022,  Aly suggested that I take out the sponsor dropdown menu so the 'company_sponsor' field will be removed.  You can read his comment below. Also, in a comment below to Aly, I explain how 'status_change_date' works, which might be helpful for you to know how that works.

alysivji commented:
Just a few comments:
- don't think the job poster needs to know all the ChiPy sponsors; can remove the dropdown. I think a checkbox is good enough. Admins can manually cross-reference
- I would move poster information to the top, you usually enter your name then enter other information
- `status_change_date` maybe we should use an audit log to capture information; this way we don't lose track of when previous changes were made
  - might be overkill for MVP, but good to start thinking about

Not sure how the User model will tie in with registration. Can folks submit without registering for a ChiPy account?

sophm256 commented:
@alysivji Thank you for your comments! 

- Yes, I can remove the sponsor dropdown. I had written some code to check if the company name matches up with a sponsor, but I can remove that too.
  
-And yes, I will move the contact info to the top. A user will have to be a registered user with ChiPy to post a job. (They do this by logging using Social Oauth using their Google account or Github account. Doing this will create a oauth related object, as well as a Django User account and UserProfile account automatically.) Because the user has to be a registered user already to post a job, I have programmed it to pre-populate the 'first_name', 'last_name', and 'email' field already if it's there.

-'status_change_date' Yeah, logging it might be worth looking into. 

This is how the 'status_change_date' works. It helps keep track of how much time can pass before the post expires.

Let's say the user submits the job post. Then the status_change_date is the date when the job post is created. When the user's job post is approved by the admin, the status_change_date becomes the date when that happened. In the job list template, I say don't show post more than 10 days ago (from the 'status_change_date').

Let's say an admin changes the 'status' field to 'extended', the 'status_change_date' gets set to that date. So this is essentially resetting the countdown of when the post should expire. So the post will get another 10 days, before it doesn't show up in the template.

When implementing it this way, I lose the ability to keep track of the 'approval date'. Let's say it's 4 days into the extended post period. The template will show this 'extended' post as posted 4 days ago, where in reality it's been up for 14 days (because 10 days in first period + 4 days in extended period). 

**Should I have fields that keep track of the 'approval_date' and 'extended_date'?**

sophm256 commented:
Having the user be a registered ChiPy is useful for a few reasons:

-Prevents someone from creating a bunch of posts in the database as spam. Even though it won't show up on the job board without approval, it will be stored in the database.

-In the future, I can make a 'job-posts-you've-created' page that shows a list of all of the postings that the user has created in the 'pending approval' stage as well as 'approved' stage. Although I'm not sure if I'll implement this, I can make it so that they can go and edit their posting as long as the post is still in the 'pending approval' stage and has not gone into the 'approved' stage.

-In the 'job-posts-you've-created' page, if the post is in the 'approved' stage, it'll just have a link to the job_post_detail page for that post. If the post is 'denied', it'll say that. (Another way to tell the user their post is approved or denied would be to send them an email.)

-Also, the user can request that their post be taken down if they've hired somebody earlier than the expiration date of the post period or if they don't like the post any more for whatever reason.

**Summary** -- this way users can: 
- edit 'pending approval' posts
- check to see if their post has been approved or denied
- request that their post be taken down ( in which the 'status' field is changed to 'archived')

sophm256 commented:
@alysivji  and @elmq0022 Also, currently, the status of a job post doesn't change to 'archived' automatically. 'archived' means the post is not meant to show up on the board any more [in other words a 'soft delete']. A job post in an 'approved' or 'extended' status will show up in the job post list until X amount of days, as coded in the job_post_list template. After that, it doesn't show up because of the 'days_elapsed' property. 

As I said before, currently, I have it so that the job post shows for only 10 days. On day 11, a job post with an 'approved' or 'extended' status will not show up any more. This is good because the job post will not show past the allotted time. But the status of a job_post will not change automatically to 'archived'. (Technically, an admin can manually set the status of a job post to 'archived' through the admin panel.)

I'm thinking of creating a custom django-admin command that checks to see if a post is past 10 days, that is if 'days_elapsed' > 10. If so, change the post's status to 'archived'. I would then use a Heroku scheduler and schedule a job each day to do execute that custom django-admin command. 

**Does this sound like a good course of action?**

elmq0022 commented:
The schedule is a fine idea, but it's better if we can get by with out it.  Maybe just give the user an editable maturity date that is no more than 30 days from present. If the post is coming due let the user add another 30 days if they want.  

I doubt anyone on the board wants to be heavily involved in managing these.

alysivji commented:
I totally understand why we use Google and GitHub. Most developers will have a GitHub account, even if they don't use it for public work. And if not, creating a Google account for personal purposes makes sense.

Folks in the talent department probably won't have a GitHub account so they'll need to use Google login. What if the job poster works at a place that doesn't have GSuite?

They'll need to create a work Google account, which just adds another barrier to submitting a job. Right now it's not quite clear how to Sign Up, we just have a `Log In` link. It might be good to add instructions about how folks can create an account and post a job.

alysivji commented:
For the status of each job posting, I think we should start as simple as possible. Right now I don't see a need for more than `Submitted`, `Approved`.

We just want a place for folks to post a job where admins can approve if it shows up on the main page or not. Once that's up, I think it makes sense to add things like `lapsed` and `extended`. Or `expired`.

We can do things manually until we have a sense of exactly what features are needed.

sophm256 commented on chipy_org/apps/job_board/views.py:
```diff
@@ -0,0 +1,57 @@
+from django.db.models import Q
+from itertools import chain
+
+@login_required
+def create_job_post(request):
+
+
+    if request.method == 'POST':
```
Thanks for the recommendation. I originally started it with trying to implement it with a ListView, but I found it to be limiting. Then I switched to a function based view since I've worked with it before and I can mold it in the future as I anticipate we'll be making changes to these views.

eevelweezel replied:
I second these recommendations.  Splitting the logic for get vs post into methods is much cleaner, and generally easier to maintain, over time.

sophm256 commented on chipy_org/apps/job_board/urls.py:
```diff
@@ -0,0 +1,14 @@
+from .views import create_job_post, after_submit_job_post, job_post_list, job_post_detail
+
+urlpatterns = [
+    url(r'^create-job-post/$', create_job_post, name='create-job-post'),
+    url(r'^job-post-list/$', job_post_list, name='job-post-list'),
+    url(r'^job-post-detail/(?P<pk>\d+)/$', job_post_detail, name='job-post-detail'),
+    url(r'^after-submit-job-post/$', after_submit_job_post, name='after-submit-job-post'),
```
Great suggestion. I'll do that.

sophm256 commented on chipy_org/apps/job_board/models.py:
```diff
@@ -0,0 +1,95 @@
+    is_sponsor = models.BooleanField(
+        default=False, verbose_name="Is the company a sponsor of ChiPy?")
+
+    company_sponsor = models.ForeignKey(Sponsor, blank=True, null=True, verbose_name="If your company is a sponsor of ChiPy, please select it from this list.")
+
```
Please see the comment I left you earlier above on this issue. I'll be removing this field because of a suggestion from Aly.

sophm256 commented on chipy_org/apps/job_board/models.py:
```diff
@@ -0,0 +1,95 @@
+    
+    description = models.CharField(max_length=2500, help_text = '2500 Character Limit')
+
+    is_sponsor = models.BooleanField(
+        default=False, verbose_name="Is the company a sponsor of ChiPy?")
```
This is a field I will keep as I'll be removing is_verified_sponsor.

sophm256 commented on chipy_org/apps/job_board/models.py:
```diff
@@ -0,0 +1,95 @@
+        else:
+            return None
+    
+    @property
+    def verify_sponsor(self):
+        # checks to see if the company_name that the user has entered is one of the sponsors
+        
+        if self.is_sponsor and self.company_sponsor:
+            company_name = self.company_name.strip().lower()
+            company_sponsor = self.company_sponsor.name.strip().lower()
+            return company_name == company_sponsor 
+        else:
+            return False
```
I will be removing this whole property/function because Aly suggested that the admins will be manually checking to see if the company is a sponsor or not if the user checks the 'is_sponsor' box.

sophm256 commented on chipy_org/apps/job_board/models.py:
```diff
@@ -1,3 +1,55 @@
 
-# Create your models here.
+MAX_LENGTH = 255
+STATUS_CHOICES = [('pending', 'Pending Approval'),
+                   ('approved', 'Approved'),
+                   ('denied', 'Denied'),
+                   ('extended', 'Extended'),
+                   ('archived', 'Archived') 
```
Yes, I can make the key shorter 👍

sophm256 commented on chipy_org/apps/job_board/models.py:
```diff
@@ -0,0 +1,95 @@
+    # - This field will change every time you update the 'status' field. Code for doing that is in the 'save' method.
+    status_change_date = models.DateTimeField(editable=False, auto_now_add=True) 
+
+    link_to_company_page = models.CharField(max_length=MAX_LENGTH)
+
```
I went with something in between and put company_website :smile:

sophm256 commented on chipy_org/apps/job_board/models.py:
```diff
@@ -0,0 +1,95 @@
+        # Query filtering and ordering in Django views are based only on fields, not properties.
+        self.is_verified_sponsor = self.verify_sponsor
+
+        return super(JobPost, self).save(*args, **kwargs)
```
We are both right. For Django 1.11 according to the documentation, it's `super(JobPost, self)` https://docs.djangoproject.com/en/1.11/topics/db/models/#overriding-model-methods but for Django 2.0 and 3.0, it's `super()` https://docs.djangoproject.com/en/2.0/topics/db/models/#overriding-model-methods

You're right about me not needing a return statement. I do get confused though where I'm supposed to be calling super in a function. Do I call it at the beginning of a method or at the end of a method or in the middle?

sophm256 replied:
@elmq0022 On the topic of super(), it seems that it's a Python version thing, and not related to the version of Django. For older versions of Python, it was super(ClassName, self). https://teamtreehouse.com/community/super-takes-at-least-1-argument-0-given

From the first answer in the Stackoverflow post, super() without arguments was introduced in Python 3.0 .  https://stackoverflow.com/questions/10482953/python-extending-with-using-super-python-3-vs-python-2

sophm256 commented on chipy_org/apps/job_board/models.py:
```diff
@@ -0,0 +1,95 @@
+        return f"{self.position} at {self.company_name}"
+
+    def __init__(self, *args, **kwargs):
+        super(JobPost, self).__init__(*args, **kwargs)
```
See above comment I wrote to you regarding super().

sophm256 commented:
@alysivji I've simplified the choices to 'submitted' and 'approved'.

sophm256 commented:
Thank you @elmq0022 and @alysivji for your helpful comments. Based on your suggestions I made changes and another commit. Below are the pictures to reflect that.

### Top half of create a job post page
I made the following changes
- simplied url to 'job/create/' (as suggested by Aaron) 
- moved contact info to the top (as suggested by Aly)
![create-post1](https://user-images.githubusercontent.com/53249319/79296172-496d7a00-7ea0-11ea-8493-dc03fec9eab4.JPG)

### Lower Half of Create a Job Post Page
Changes I made:
- removed the 'company_sponsor' field (as suggested by Aly)
- changed label of field 'link_to_company_page' to 'company_website'
![create-post2](https://user-images.githubusercontent.com/53249319/79296181-4d010100-7ea0-11ea-9808-4ca90f90da93.JPG)

### One job post in Django Admin
Changes I made:

- Added the `'days_to_expire'` field that shows up here with the verbose label as `'num of days for post to show'`. The default for this field is 30 when a job post is created. That is, in 30 days from the approval date, the post will expire and not show up any more. This field can be editable by the admin if the admin wants to give the person more than or less than 30 days. 

- Also changed the 'agree_to_terms' labeling to be consistent with what shows up in the create a job post form

![admin-single-job-post](https://user-images.githubusercontent.com/53249319/79296197-5d18e080-7ea0-11ea-8b99-523dbf01d942.JPG)

### The list of Job Posts in Django Admin
**If you click on the picture below, I think it should get bigger so you can see the words better on your screen.**

I added two non editable fields:` 'approval_date'` and `'expiration_date'` .  These two fields are only viewable from this list view. Initially the 'approval_date' and 'expiration_date' is empty when job post is first created. When the status is changed to 'approved', the 'approval_date' is set to that date. Also, the 'expiration_date' is also set by adding number of days from the 'days_to_expire' field to the 'approval_date' field.  The 'expiration_date' will be used to to filter what posts show up in the job board.

At Aly's suggestion, I simplified the 'status' choices to be 'submitted' and 'approved'. I took out the 'status_change_date' field from this view but the code still remains in case we need to use it later. But for now, I'm not depending on that field for anything.

**Note: If a post is approved and has the `'days_to_expire'` field (which shows up here with the verbose label `'num_of_days_for_post_to_show'`) manually set to zero or a negative number, the post will not show up on the job board. This may be useful if an admin wants to signal that post had been approved and was up for awhile, but has been deactivated. In this case, we may not need to have a status called 'archived'.** 

Also, it's possible to delete the job post from the admin manually.


![admin-list](https://user-images.githubusercontent.com/53249319/79296205-61dd9480-7ea0-11ea-87d6-9164d8819769.JPG)

### Job List Page
This page really hasn't changed except for a more simplified url. I only include it here so that you can see what job posts do show up given the settings from the Django Admin job list that I had on the previous picture.
![list](https://user-images.githubusercontent.com/53249319/79296219-6c982980-7ea0-11ea-9199-bbccac103182.JPG)

sophm256 commented on chipy_org/apps/job_board/models.py:
```diff
@@ -0,0 +1,95 @@
+    def days_elapsed(self):
+
+        # computes days elapsed from when job post is put in 'approved' or 'extended' status
+        if self.status == 'approved' or self.status == 'extended':
```
Yes, I've removed 'extended'. I also created an 'expiration_date' field, as well as a 'days_to_expire' field. I explain what these are in the pictures below that I posted after I made another commit.

alysivji replied:
Let's not worry about expired jobs just yet. I think we're losing track that there are currently 0 jobs on our jobs board and we're still figuring out what needs to get built.

My experience has taught me that you should get the most basic MVP out the door and into the hands of customers / users. Only when people use your software, you understand what you need to build.

As long as we have a `created_at` and `modified_at` field, we have up to 60 days since this goes live to write up the expiration feature.

sophm256 replied:
@alysivji Thanks for your feed back. I have coded in the expired feature already. It currently defaults to 30 days, but I can change it to 60 days in my next commit, which probably will happen sometime soon. I'm going to keep this feature in there if that's okay with you. If for some reason you don't like the way it's working, we can revisit it in the future and I can rework it. 

Also, I asked @eevelweezel to take a look at my PR and depending on her comments, I might make more changes.

eevelweezel replied:
I like the idea of allowing this to be user-editable, and keeping it as simple as possible. This functionality sounds reasonable, for the time being.

elmq0022 commented:
That's a fair point.  And it should be an easy add down the line.

I think it has already been added. Personally I would not take it out, but
I will step back let you and sophan decide.  Thanks!

On Wed, Apr 15, 2020, 9:04 AM Aly Sivji <notifications@github.com> wrote:

> *@alysivji* commented on this pull request.
> ------------------------------
>
> In chipy_org/apps/job_board/models.py
> <https://github.com/chicagopython/chipy.org/pull/271#discussion_r408867798>
> :
>
> > +        if  self.__original_status != self.status:
> +            self.status_change_date = datetime.now()
> +            self.__original_status = self. status
> +
> +        # The property 'verify_sponsor' is saved in a model field because
> +        # properties can not be used in filtering querysets.
> +        # Query filtering and ordering in Django views are based only on fields, not properties.
> +        self.is_verified_sponsor = self.verify_sponsor
> +
> +        return super(JobPost, self).save(*args, **kwargs)
> +
> +    @property
> +    def days_elapsed(self):
> +
> +        # computes days elapsed from when job post is put in 'approved' or 'extended' status
> +        if self.status == 'approved' or self.status == 'extended':
>
> Let's not worry about expired jobs just yet. I think we're losing track
> that there are currently 0 jobs on our jobs board and we're still figuring
> out what needs to get built.
>
> My experience has taught me that you should get the most basic MVP out the
> door and into the hands of customers / users. Only when people use your
> software, you understand what you need to build.
>
> As long as we have a created_at and modified_at field, we have up to 60
> days since this goes live to write up the expiration feature.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/chicagopython/chipy.org/pull/271#discussion_r408867798>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA2KPYB5NAH4K2U7NADIGHLRMW5G7ANCNFSM4MGWCM3A>
> .
>

raymondberg commented on chipy_org/apps/job_board/templates/job_post_list.html:
```diff
@@ -0,0 +1,35 @@
+                {% if post.is_sponsor %}
+                    <h3>Sponsored Job Posts</h3>
+                {% else %}
+                    <h3>Regular Job Posts</h3>
```
Can we change this to `Other Job Posts`?

raymondberg commented on chipy_org/apps/job_board/templates/job_post_list.html:
```diff
@@ -0,0 +1,35 @@
+                {% elif post.days_elapsed == 1 %}
+                    1 day ago 
+                {% else %}
+                    {{post.days_elapsed}} days ago
```
For the second two conditions you can switch to `pluralize`.

```suggestion
                    {{post.days_elapsed}} day{{ post.days_elapsed|pluralize }} ago
```

https://docs.djangoproject.com/en/3.0/ref/templates/builtins/#pluralize

elmq0022 commented:
One other thing this did pull requests was generating lint errors last I looked at it.  Easiest thing to do might be to use something like autopep8 or black to clean it up a bit.  Just try to run the tool on your directory only.
